src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java

```diff
@@ -0,0 +1,81 @@
+package org.veupathdb.service.eda.ss.model.reducer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.gusdb.fgputil.functional.Functions;
+import org.gusdb.fgputil.functional.TreeNode;
+import org.veupathdb.service.eda.ss.model.Entity;
+import org.veupathdb.service.eda.ss.model.Study;
+import org.veupathdb.service.eda.ss.model.filter.Filter;
+import org.veupathdb.service.eda.ss.model.filter.SingleValueFilter;
+import org.veupathdb.service.eda.ss.model.reducer.ancestor.EntityIdIndexIteratorConverter;
+import org.veupathdb.service.eda.ss.model.variable.Variable;
+import org.veupathdb.service.eda.ss.model.variable.VariableValueIdPair;
+import org.veupathdb.service.eda.ss.model.variable.VariableWithValues;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Constructs data flow trees for tabular subsetting requests. These trees represent the flow of data for a particular
+ * request. The tree will be rooted with the entity for which we are outputting one or more variables.
+ */
+public class DataFlowTreeFactory {
+  private static final Logger LOG = LogManager.getLogger(DataFlowTreeFactory.class);
+
+  public TreeNode<DataFlowNodeContents> create(TreeNode<Entity> prunedEntityTree,
+                                               Entity outputEntity,
+                                               List<Filter> filters,
+                                               Study study) {
+    final TreeNode<Entity> outputNode = prunedEntityTree.findFirst(entity -> entity.equals(outputEntity));
+    // In lieu of a pointer up the tree, we use this function to traverse from the original root to a node's parent.
+    final Function<TreeNode<Entity>, TreeNode<Entity>> parentMapper = child ->
```

Comment:
I wonder if you want to return an Optional here vs null for the root entity.  Could enable later logic like parent.ifPresent(parent -> /*do something*/) + NPE prevention.

Reply:
Makes sense, addressed this in coming PR.